File: content/cpp-driver/upcoming/source/includes/usage-examples/index-code-examples.cpp
/*
To build:
brew install mongo-cxx-driver
clang++ -o indexes main.cpp -std=c++17 $(pkg-config --libs --cflags libmongocxx)
*/
#include <iostream>

#include <bsoncxx/builder/basic/document.hpp>
#include <bsoncxx/json.hpp>
#include <mongocxx/client.hpp>
#include <mongocxx/instance.hpp>
#include <mongocxx/uri.hpp>

using bsoncxx::builder::basic::kvp;
using bsoncxx::builder::basic::make_document;

int main() {

    mongocxx::instance instance;
    mongocxx::uri uri("<connectionString>");
    mongocxx::client client(uri);

    auto db = client["<databaseName>"];
    auto collection = db["<collectionName>"];
    
    {
        // start-single-field
        auto index_specification = make_document(kvp("<fieldName>", 1));
        auto result = collection.create_index(index_specification.view());

        std::cout << "Index created: " << bsoncxx::to_json(result) << std::endl;
        // end-single-field
    }
    {
        // start-compound-field
        auto index_specification = make_document(kvp("<fieldName1>", -1), kvp("<fieldName2>", -1));
        auto result = collection.create_index(index_specification.view());

        std::cout << "Index created: " << bsoncxx::to_json(result) << std::endl;
        // end-compound-field
    }
    {
        // start-remove-index
        collection.indexes().drop_one("<indexName>");

        std::cout << "Index dropped." << std::endl;
        // end-remove-index

        // start-remove-all-indexes
        collection.indexes().drop_all();

        std::cout << "All indexes removed." << std::endl;
        // end-remove-all-indexes
    }
    {
        // start-create-search-index
        // Create an index model with your index name and definition
        auto siv = collection.search_indexes();
        auto name = "<searchIndexName>";
        auto definition = make_document(kvp("mappings", make_document(kvp("dynamic", true))));
        auto model = mongocxx::search_index_model(name, definition.view());
    
        // Create the search index
        auto result = siv.create_one(model);
        std::cout << "New index name: " << result << std::endl;
        // end-create-search-index
    }
    {    
        // start-list-search-indexes
        auto siv = collection.search_indexes();
        auto result = siv.list(); 
        for (const auto &idx : result) {
            std::cout << bsoncxx::to_json(idx) << std::endl;
        }
        // end-list-search-indexes
    }
    {
        // start-update-search-index
        auto siv = collection.search_indexes();
        auto update_fields = make_document(kvp("<fieldName>", make_document(kvp("type", "<fieldType>"))));
        auto update_definition = make_document(kvp("mappings", make_document(kvp("dynamic", false), kvp("fields", update_fields))));
        siv.update_one("<searchIndexName>", update_definition.view());
        // end-update-search-index   
    }
    {
        // start-remove-search-index
        auto siv = collection.search_indexes();
        siv.drop_one("<searchIndexName>");
        // end-remove-search-index
    }
}

File: current/source/includes/indexes/index-code-examples.rb
require 'mongo'

# Replace the placeholders with your credentials
uri = "<connection string>"

# Sets the server_api field of the options object to Stable API version 1
options = { server_api: { version: '1' }}

# Creates a new client and connects to the server
client = Mongo::Client.new(uri, options)

database = client.use('<database name>')
collection = database[:<collection name>]

# Single Field
# start-index-single
collection.indexes.create_one({ <field name>: 1 })
# end-index-single

# Compound
# start-index-compound
collection.indexes.create_one({ <field name 1>: -1, <field name 2>: 1 })
# end-index-compound

# Multikey
# start-index-multikey
collection.indexes.create_one({ <field name>: 1 })
# end-index-multikey

# Geospatial
# start-index-geospatial
collection.indexes.create_one({ <GeoJSON field name>: '2dsphere' })
# end-index-geospatial

# Atlas Search

# Create Search Index
# start-create-search-index
index_definition = {
  mappings: {
    dynamic: false,  
    fields: { 
      <field name>: { type: '<field type>' }
    }
  }
}
collection.search_indexes.create_one(index_definition, name: '<index name>')
# end-create-search-index

# List Search Indexes
# start-list-search-indexes
puts collection.search_indexes.collect(&:to_json)
# end-list-search-indexes

# Update Search Indexes
#start-update-search-indexes
updated_definition = {
  mappings: {
    dynamic: false,  
    fields: { <updated field name>: { type: '<updated field type>' } }
    }
}

collection.search_indexes.update_one(updated_definition, name: '<index name>')
#end-update-search-indexes

# Delete Search Index
# start-drop-search-index
collection.search_indexes.drop_one(name: '<index name>')
# end-drop-search-index

# Text Index
# start-text
collection.indexes.create_one({ <field name>: 'text' })
# end-text

# Create Many
# start-index-create-many
collection.indexes.create_many([
  { key: { <field name 1>: 1 } },
  { key: { <field name 2>: -1 } },
])
# end-index-create-many

# Delete an Index
# start-drop-single-index
collection.indexes.drop_one( '<index name>' )
# end-drop-single-index

# Drops all indexes in the collection.
# start-drop-all-index
collection.indexes.drop_all
# end-drop-all-index

# List an Index
# start-list-indexes
puts collection.indexes.collect(&:to_json)
# end-list-indexes

client.close
